### ISSUE TITLE
[BUG] _CMAKE build on macOS fails with KIM-API build with clang (cc, c++) and gfortran_

### ISSUE THREAD

**Issue by ellio167** [lammps/lammps#1439]
**Summary**

CMake build on macOS with KIM-API and cc and c++ (clang), gfortran fails with cannot find library 'gfortran'

**LAMMPS Version and Platform**

patch_30Apr2019

macOS Sierra (10.12.6)

cc -v
Apple LLVM version 9.0.0 (clang-900.0.39.2)
Target: x86_64-apple-darwin16.7.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin

c++ -v
Apple LLVM version 9.0.0 (clang-900.0.39.2)
Target: x86_64-apple-darwin16.7.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin

gfortran -v
Using built-in specs.
COLLECT_GCC=gfortran
COLLECT_LTO_WRAPPER=/usr/local/Cellar/gcc/8.3.0/libexec/gcc/x86_64-apple-darwin16.7.0/8.3.0/lto-wrapper
Target: x86_64-apple-darwin16.7.0
Configured with: ../configure --build=x86_64-apple-darwin16.7.0 --prefix=/usr/local/Cellar/gcc/8.3.0 --libdir=/usr/local/Cellar/gcc/8.3.0/lib/gcc/8 --disable-nls --enable-checking=release --enable-languages=c,c++,objc,obj-c++,fortran --program-suffix=-8 --with-gmp=/usr/local/opt/gmp --with-mpfr=/usr/local/opt/mpfr --with-mpc=/usr/local/opt/libmpc --with-isl=/usr/local/opt/isl --with-system-zlib --with-pkgversion='Homebrew GCC 8.3.0' --with-bugurl=https://github.com/Homebrew/homebrew-core/issues
Thread model: posix
gcc version 8.3.0 (Homebrew GCC 8.3.0)

**Expected Behavior**

Build succeeds in linking.

**Actual Behavior**

...
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++  -O3 -DNDEBUG -Wl,-search_paths_first -Wl,-headerpad_max_install_names  CMakeFi
ld: library not found for -lgfortran
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [lmp] Error 1
make[1]: *** [CMakeFiles/lmp.dir/all] Error 2

**Steps to Reproduce**

mkdir build
cd build
( FC=gfortran-8 cmake ../cmake/ -DCMAKE_VERBOSE_MAKEFILE:BOOL=ON -DPKG_KIM=ON -DDOWNLOAD_KIM=ON && make ) > configure-and-build.log 2>&1

See attached configure & build log for output.

[configure-and-build.log](https://github.com/lammps/lammps/files/3133647/configure-and-build.log)

**Comment by akohlmey** [comment]
@ellio167 i am not certain, if this needs to be addressed on the LAMMPS side, because there is no fortran code in LAMMPS itself, but in the KIM-API library files. Those should be linked with libgfortran and whatever else is needed and not LAMMPS. 

Also, on Linux, when installing a custom compiler (available via environment module) we add to the environment variables `LIBRARY_PATH` and `LD_RUN_PATH` the path to the runtime libraries of the custom compiler. The first helps at link time to find the library, the second embeds the run time path into the executable or shared object/library (similar to setting LD_LIBRARY_PATH). I don't know what the MacOSX equivalent of those variables is.

**Comment by ellio167** [comment]
@akohlmey The KIM API library is linked with gfortran and all is fine from that perspective.  The lammps executable should only need to be linked with libkim-api, it should not need to explicitly link against libgfortran.

However, somehow the LAMMPS build is __erroneously__ adding the "-lgfortran" flag to its link command.  This has something to do with the LAMMPS CMake setup.  I'm pretty sure the link would succeed if simply the "-lgfortran" were removed.

**Comment by akohlmey** [comment]
@ellio167 i think i understand, where this is coming from. When downloading KIM-API, the CMakeLists.txt file enables Fortran compilation, so that the same fortran compiler is used in KIM-API compilation as is used elsewhere when compiling fortran code in LAMMPS. Please try the following patch.

```
[akohlmey@zero build-kim]$ git diff
  diff --git a/cmake/CMakeLists.txt b/cmake/CMakeLists.txt
  index 63bc16827..0791d61cb 100644
  --- a/cmake/CMakeLists.txt
  +++ b/cmake/CMakeLists.txt
  @@ -735,7 +735,6 @@ if(PKG_KIM)
     if(DOWNLOAD_KIM)
       message(STATUS "KIM-API download requested - we will build our own")
       enable_language(C)
  -    enable_language(Fortran)
       include(ExternalProject)
       ExternalProject_Add(kim_build
         URL https://s3.openkim.org/kim-api/kim-api-2.0.2.txz
  @@ -743,7 +742,6 @@ if(PKG_KIM)
         BINARY_DIR build
         CMAKE_ARGS -DCMAKE_C_COMPILER=${CMAKE_C_COMPILER}
                    -DCMAKE_CXX_COMPILER=${CMAKE_CXX_COMPILER}
  -                 -DCMAKE_Fortran_COMPILER=${CMAKE_Fortran_COMPILER}
                    -DCMAKE_INSTALL_PREFIX=<INSTALL_DIR>
                    -DCMAKE_BUILD_TYPE=${CMAKE_BUILD_TYPE}
         )
```

Apart from that, I still believe that you would have to set `LIBRARY_PATH` and one of `LD_RUN_PATH` or `LD_LIBRARY_PATH` in order to link a c++ executable linked to fortran code in LAMMPS, e.g. from lib/linalg/liblinalg.a when using a fortran compiler with its runtime in a non-standard location.

**Comment by ellio167** [comment]
The suggested patch essentially works, but it needed tweaking to remember any compiler environment variable settings...  I'll submit a PR right away.

I tested it, and there is no need for a `LIBRARY_PATH` or other variable setting at run time.  The full absolute kim-api library path is embedded in the lammps executable (I think this is true on linux too, at least for the kim-api library).

**Comment by akohlmey** [comment]
@ellio167 would you mind trying to test compiling LAMMPS with LATTE and USER-AWPMD enabled as well. That should enable fortran support and thus may fail linking on your MacOSX box as well. If it works, i would like to see the cmake configure output. perhaps we can gain some idea for how to address the handling of fortran with your kind of setup in a more generic way.

**Comment by ellio167** [comment]
@akohlmey It tried it (using the code in this PR) and it, indeed, failed.  In fact, if you look at the log, you will see that kim-api is compiled with `gfortran-8` but LATTE is compiled with `pgf95`
[cmake-make.log](https://github.com/lammps/lammps/files/3140392/cmake-make.log)

Since this is creating a static library, when linking lammps both liblatte.a and the fortran library will need to be listed on the link line.  I think, maybe, the cmake way to do this would be to create an "imported library" and then set the "LINKER_LANGUAGE" to Fortran for the imported library...  maybe...

**Comment by akohlmey** [comment]
> @akohlmey It tried it (using the code in this PR) and it, indeed, failed. In fact, if you look at the log, you will see that kim-api is compiled with `gfortran-8` but LATTE is compiled with `pgf95`
> [cmake-make.log](https://github.com/lammps/lammps/files/3140392/cmake-make.log)

thanks for taking the time. this shows, that for LATTE, the cmake compiler settings are not passed on as it is for other external libraries. this was the idea also for KIM-API to define the CMAKE_XXX_COMPILER variables on the cmake command line and then use them for LAMMPS and all downloaded libraries as well. all things considered, i like the original setup with the unneeded -lgfortran better than your PR, as that would allow us to keep compilers consistent more easily. like i mentioned before, it should be possible to work around the library not found error with an environment variable or an extra linker flag passed to CMake. particularly with fortran, it should be avoided to have objects or libraries compiled with different compilers linked together.

> Since this is creating a static library, when linking lammps both liblatte.a and the fortran library will need to be listed on the link line. I think, maybe, the cmake way to do this would be to create an "imported library" and then set the "LINKER_LANGUAGE" to Fortran for the imported library... maybe...

this is beyond my level of knowledge in cmake.

**Comment by ellio167** [comment]
@akohlmey @rbberger @junghans 
Following on....  I think I've worked out a clean way to handle these sorts of external libraries.  It might serve as a template for lammps cmake support.  However, I understand if there is not interest in the large refactoring this would lead to....

Here is my test project:
[latte-external-package-build.zip](https://github.com/lammps/lammps/files/3142309/latte-external-package-build.zip)

which contains this file
[CMakeLists.txt](https://github.com/lammps/lammps/files/3142310/CMakeLists.txt)

and a simple c++ file that calls the latte_abiversion() routine and prints the reuslt.

**Comment by akohlmey** [comment]
@ellio167 thanks for your efforts. I've adapted passing additional settings to the LATTE library build for improved consistency into PR #1450 
Since we want to do a stable release very soon, i would suggest we hold off on larger changes to the cmake configuration scripts until after the stable release and otherwise try. At the moment, things seem to be working ok for the more common case of Linux machines and i don't want to risk breaking that in unexpected ways while trying to fix it for other platforms.

**Comment by ellio167** [comment]
@akohlmey Makes sense.

**Comment by akohlmey** [comment]
@ellio167 just found a simple solution for the KIM problem on MacOSX (and possibly other platforms), where we want the fortran compiler setting to be consistent, but not automatically link with it directly.
This can be done with the `check_language()` function instead of `enable_language()`. This also sets `${CMAKE_Fortran_COMPILER}` and thus allows to set it globally to be consistent, but does not add `-lgfortran` to the linker command line, and hence no more failures on MacOSX out of the box. see the following commit, that will hopefully be merged tomorrow as part of PR #1604
5b4b418289c86fbed8a4fc93b72633e1cce304b9

**Comment by ellio167** [comment]
@akohlmey Nice!  I've confirmed that this works as described.
